### DISCUSSION TITLE
Problem using dfu-util to install FlashFloppy

### DISCUSSION THREAD

**Discussion by AlpacaLips99** [keirf/flashfloppy#639] [Q&A]
Hoping for help with installation of Gotek version SFRC922 (older STM32F105 processor). I'm using a MacBookPro (Intel) with MacOS 12.3 (Monterey). I installed dfu-util using MacPorts. Connecting to Gotek via Thunderbolt to USB-A dongle and a USB-A to USB-A cable. Although I got my Gotek from a private party, I was told that it was new old stock and had never been flashed. I'm attempting to prepare the Gotek to use FlashFloppy for use in an Amiga 500. I used drag/drop to ensure the path to the .dfu is correct. I'm attempting to use the instructions at https://github.com/keirf/flashfloppy/wiki/Firmware-Programming. 

First command to reset the Gotek:

> $ sudo dfu-util -a 0 -s :unprotect:force -D /Users/username/Downloads/flashfloppy-v3.30/FF_Gotek-v3.30.dfu 
> Password:
> dfu-util 0.11
> Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
> Copyright 2010-2021 Tormod Volden and Stefan Schmidt
> This program is Free Software and has ABSOLUTELY NO WARRANTY
> Please report bugs to http://sourceforge.net/p/dfu-util/tickets/
> Match vendor ID from file: 0483
> Match product ID from file: df11
> Multiple alternate interfaces for DfuSe file
> Opening DFU capable USB device...
> Device ID 0483:df11
> Device DFU version 011a
> Claiming USB DFU Interface...
> Setting Alternate Interface #0 ...
> Determining device status...
> DFU state(4) = dfuDNBUSY, status(0) = No error condition is present
> DFU mode device DFU version 011a
> Device returned transfer size 2048
> DfuSe interface name: "Internal Flash "
> dfu-util: dfuse_download: libusb_control_transfer returned -9 (LIBUSB_ERROR_PIPE)
> dfu-util: Error during special command "READ_UNPROTECT" download

Second command to install FlashFloppy firmware:

> $ sudo dfu-util -a 0 -D /Users/username/Downloads/flashfloppy-v3.30/FF_Gotek-v3.30.dfu
> dfu-util 0.11
> Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
> Copyright 2010-2021 Tormod Volden and Stefan Schmidt
> This program is Free Software and has ABSOLUTELY NO WARRANTY
> Please report bugs to http://sourceforge.net/p/dfu-util/tickets/
> Match vendor ID from file: 0483
> Match product ID from file: df11
> Multiple alternate interfaces for DfuSe file
> Opening DFU capable USB device...
> Device ID 0483:df11
> Device DFU version 011a
> Claiming USB DFU Interface...
> Setting Alternate Interface #0 ...
> Determining device status...
> DFU state(10) = dfuERROR, status(15) = Device stalled an unexpected request
> Clearing status
> Determining device status...
> DFU state(2) = dfuIDLE, status(0) = No error condition is present
> DFU mode device DFU version 011a
> Device returned transfer size 2048
> DfuSe interface name: "Internal Flash "
> File contains 1 DFU images
> Parsing DFU image 1
> Target name: ST...
> Image for alternate setting 0, (2 elements, total size = 113032)
> Setting Alternate Interface #0 ...
> Parsing element 1, address = 0x08000000, size = 29760
> Erase [ ] 0% 0 bytesdfu-util: dfuse_download: libusb_control_transfer returned -7 (LIBUSB_ERROR_TIMEOUT)
> dfu-util: Error during special command "ERASE_PAGE" download

After executing these commands, I removed the jumpers and installed in in my Amiga 500. On startup, the Gotek's green light comes on momentarily and the 7-segment LED display just says "- - -".

I see that there are multiple errors here, but it isn't clear to me which ones (if not all) are actually problematic, or how to resolve the issue. I hope the problem will be obvious to someone more knowledgeable.

Does anyone see where I'm going wrong?

**Comment by keirf** [comment]
This Gotek is running Cortex Amiga firmware. Not that it matters. The DFU bootloader will still work.

The libusb errors at the end of each command are the problem. It could be related to connecting via a thunderbolt adapter (should work but most people will be using a simple usb2 a-a cable). It could be dfu utils version. Or something else unfortunately. The simplest method is ST software, but on Windows unfortunately.

**Comment by AlpacaLips99** [comment]
Thanks for your quick response. So, if it's already running Cortex (not sure how you can tell that), do I need FlashFloppy at all?

Since Macs have no USB-A ports and require the installation of Apple developer tools (and thus registration as a developer with Apple) just to obtain the prerequisite components for dfu-util, I definitely can't recommend Gotek as a solution for DIY'ers who rely on Macs. (In other words, non-Windows users should just pay the money to buy a Gotek already set up for their particualar needs. It would be nice if there were a Mac app with a GUI to set up a Gotek.

I have a Windows 8 box buried somewhere that I could dig out to get this thing done. What do you mean by "ST software"?

**Comment by keirf** [comment]
Cortex is the only firmware I know that displays `---`.

Surprised if dfu-util needs more than Xcode command line tools -- do they also need developer registration? `xcode-select --install` in macOS Terminal? Anyway, I suppose it is moot if it doesn't work.

Best STM software is the Cube programmer: https://www.st.com/en/development-tools/stm32cubeprog.html

**Reply by keirf** [reply]
And look at that, there's a version for Mac too!

**Comment by keirf** [comment]
I also see in the FlashFloppy wiki: "There are reports of some issues connecting as a DFU device to macOS. In some cases this can be solved by connecting via an external USB hub."

So that might be worth investigating too!

**Reply by AlpacaLips99** [reply]
I had to register as a developer to get Xcode, and then installing the command line tools was an extra step in Terminal. 

I saw that note about connecting via a USB hub; that's basically what I'm doing since a direct USB-A connection is impossible. It's one of those thunderbolt Swiss army knife dongles that does USB-A, HDMI, Ethernet, etc. I'm assuming it's essentially the same as a powered USB hub, because it has a power pass-through that charges the Mac and USB devices plugged into it.

I'm going to give STM32CubeProgrammer a try on the Mac, and if that doesn't work I'll dig out the Windows 8 box and try the Win32 version. And if that doesn't work then I guess Gotek isn't for me.

**Reply by AlpacaLips99** [reply]
When I use STM32CubeProgrammer, I need to tell it to use the FF_Gotek-v3.30.dfu firmware, correct?

**Reply by keirf** [reply]
Yes that's right.

**Comment by AlpacaLips99** [comment]
Thanks again!

**Comment by AlpacaLips99** [comment]
STM32CubeProgrammer for Mac is buggier than a Kenyan termite castle, and they didn't bother writing Mac support docs. I'll try Windows 8 options this weekend if I can find my dusty old box, and if that doesn't work then I'm giving up on Gotek programming flashing. I've wasted enough time on this.

**Comment by AlpacaLips99** [comment]
I've been doing a ton of tinkering without success but I have some clues. I've tried the ST software (3 different apps), mostly on Windows 7 x86, with no USB hub. This was difficult because I have no way to connect this computer to the internet (drivers!). I also tried a lot of seemingly random things like different USB ports, trying both USB and USB-TTL, attempting to reset the Gotek at various points by holding down the right button at powerup, briefly jumpering RST and GND, and alternate jumper settngs described at https://cygnusalpha.blogspot.com/2018/10/getting-flashfloppy-onto-gotek-sfrc922d.html (even though my model is SFRC922, and not SFRC922D). There seems to be no end to the mishmash of advice on this online, to the point of starting to look like folksy home remedies. The only promising thing that happened was that when I ran DFU Tester 3.0.1 in Windows (part of STM's Dfuse v3.0.6 package), I was able to "GetState" and "GetStatus", select the .dfu to download, and execute the download. A progress bar ran (~10 seconds), and returned "Success," yet the Gotek still shows no signs of life when powered up in an Amiga 500. I should mention that I had connected the temporary jumper wires described in this github site, and removed them when attempting to run the Gotek in the Amiga. 

I reconnected the temporary jumpers and reconnected to my Mac vis powered USB hub (Thunderbolt adapter) and A-A cable. The ST software for Mac was buggy, inscrutible, and undocumented, so I returned to dfu-util in Terminal. When attempting dfu-util -D, I now get this:

```
$ dfu-util -a 0 -D /Users/username/Downloads/flashfloppy-v3.30/FF_Gotek-v3.30.dfu
dfu-util 0.11

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2021 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Match vendor ID from file: 0483
Match product ID from file: df11
Multiple alternate interfaces for DfuSe file
Opening DFU capable USB device...
Device ID 0483:df11
Device DFU version 011a
Claiming USB DFU Interface...
Setting Alternate Interface #0 ...
Determining device status...
DFU state(10) = dfuERROR, status(10) = Device's firmware is corrupt. It cannot return to run-time (non-DFU) operations
Clearing status
Determining device status...
DFU state(2) = dfuIDLE, status(0) = No error condition is present
DFU mode device DFU version 011a
Device returned transfer size 2048
DfuSe interface name: "Internal Flash  "
File contains 1 DFU images
Parsing DFU image 1
Target name: ST...
Image for alternate setting 0, (2 elements, total size = 113032)
Setting Alternate Interface #0 ...
Parsing element 1, address = 0x08000000, size = 29760
Erase   	[                         ]   0%            0 bytesdfu-util: dfuse_download: libusb_control_transfer returned -7 (LIBUSB_ERROR_TIMEOUT)
dfu-util: Error during special command "ERASE_PAGE" download
```

The part I found most interesting is this:
`DFU state(10) = dfuERROR, status(10) = Device's firmware is corrupt. It cannot return to run-time (non-DFU) operations
`

This seems to indicate that the Gotek is in DFU boot mode due to firmware corruption. 

--list really surprised me as well:
```
$ dfu-util --list
dfu-util 0.11

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2021 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Found DFU: [0483:df11] ver=2200, devnum=14, cfg=1, intf=0, path="20-1.2.4", alt=1, name="@Option Bytes  /0x1FFFF800/01*016 e", serial="STM32"
Found DFU: [0483:df11] ver=2200, devnum=14, cfg=1, intf=0, path="20-1.2.4", alt=0, name="@Internal Flash  /0x08000000/128*002Kg", serial="STM32"
```

It seems to be seeing two instances of the Gotek, which I would guess is related to the corrupt firmware.

Since -s unprotect:force doesn't seem to be getting the job done, I'm tempted to try -s mass-erase:force in hopes of returning to a clean slate that I can then download the FlashFoppy firmware onto, but I'm concerned that doing so may brick the Gotek.

At the risk of wearing out my welcome, does any of this point to a possible solution? Should I go ahead and try mass-erase? I'm very close to concluding that this Gotek is unsalvagable, and just buying one already set up for Amiga. That would be painful but I'm not getting any younger.

**Comment by AlpacaLips99** [comment]
HOLD EVERYTHING!... In one last ditch effort to try to get STM32CubeProgrammer to work, I found that it was in fact able to read the Gotek on the Mac (not sure why it couldn't before), and surprisingly it was able to download the firmware to the Gotek. I didn't selet "full chip erase" for the same reason I was worried about trying mass-erase in dfi-util; I also did **not** check "Skip flash erase before programming" though. After powering it up in the Amiga, it is working! 

I think I'll leave my previous lengthy message up for posterity.
